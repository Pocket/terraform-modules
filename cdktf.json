{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
