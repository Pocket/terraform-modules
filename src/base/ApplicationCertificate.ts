import { AcmCertificate } from '@cdktf/provider-aws/lib/acm-certificate';
import { AcmCertificateValidation } from '@cdktf/provider-aws/lib/acm-certificate-validation';
import { DataAwsRoute53Zone } from '@cdktf/provider-aws/lib/data-aws-route53-zone';
import { Route53Record } from '@cdktf/provider-aws/lib/route53-record';
import { TerraformMetaArguments, TerraformProvider } from 'cdktf';
import { Construct } from 'constructs';

export interface ApplicationCertificateProps extends TerraformMetaArguments {
  domain: string;
  /**
   * If zoneId is not passed then we use a data block and the zoneDomain to grab it.
   */
  zoneId?: string;
  zoneDomain?: string;
  tags?: { [key: string]: string };
}

/**
 * Generates an Application Certificate given a domain name and zoneId
 */
export class ApplicationCertificate extends Construct {
  public readonly arn: string;
  // Use `certificateValidation` in `dependsOn` to block on the
  // complete certificate for any downstream dependencies
  public readonly certificateValidation: AcmCertificateValidation;

  constructor(
    scope: Construct,
    name: string,
    config: ApplicationCertificateProps
  ) {
    super(scope, name);

    if (!config.zoneId && config.zoneDomain) {
      const route53Zone = new DataAwsRoute53Zone(this, `zone`, {
        name: config.zoneDomain,
        provider: config.provider,
      });
      config.zoneId = route53Zone.zoneId;
    } else if (!config.zoneId && !config.zoneDomain) {
      throw new Error('You need to pass either a zone id or a zone domain');
    }

    const certificate = ApplicationCertificate.generateAcmCertificate(
      this,
      config.domain,
      config.tags,
      config.provider
    );

    const certificateRecord = ApplicationCertificate.generateRoute53Record(
      this,
      config.zoneId,
      certificate,
      config.provider
    );

    const validation = ApplicationCertificate.generateAcmCertificateValidation(
      this,
      certificate,
      certificateRecord,
      config.provider
    );

    this.arn = certificate.arn;
    this.certificateValidation = validation;
  }

  static generateAcmCertificate(
    scope: Construct,
    domain: string,
    tags?: { [key: string]: string },
    provider?: TerraformProvider
  ): AcmCertificate {
    return new AcmCertificate(scope, `certificate`, {
      domainName: domain,
      validationMethod: 'DNS',
      tags: tags,
      lifecycle: {
        createBeforeDestroy: true,
      },
      provider: provider,
    });
  }

  static generateRoute53Record(
    scope: Construct,
    zoneId: string,
    cert: AcmCertificate,
    provider?: TerraformProvider
  ): Route53Record {
    const record = new Route53Record(scope, `certificate_record`, {
      name: cert.domainValidationOptions.get(0).resourceRecordName,
      type: cert.domainValidationOptions.get(0).resourceRecordType,
      zoneId,
      records: [cert.domainValidationOptions.get(0).resourceRecordValue],
      ttl: 60,
      dependsOn: [cert],
      provider: provider,
    });

    return record;
  }

  static generateAcmCertificateValidation(
    scope: Construct,
    cert: AcmCertificate,
    record: Route53Record,
    provider?: TerraformProvider
  ): AcmCertificateValidation {
    return new AcmCertificateValidation(scope, `certificate_validation`, {
      certificateArn: cert.arn,
      validationRecordFqdns: [record.fqdn],
      dependsOn: [record, cert],
      provider: provider,
    });
  }
}
