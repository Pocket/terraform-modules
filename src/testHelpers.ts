import { Resource, TerraformStack } from 'cdktf';
import { Construct } from 'constructs';

// used to test individual static functions within component classes
// generates a basic/empty construct Resource to provide context
export class TestResource extends Resource {
  constructor(scope: TerraformStack | Construct, name: string) {
    super(scope, name);
  }
}
