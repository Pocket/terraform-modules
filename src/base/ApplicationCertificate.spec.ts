import { Testing, TerraformStack } from 'cdktf';
import { TestResource } from '../testHelpers';
import { ApplicationCertificate } from './ApplicationCertificate';

describe('ApplicationCertificate', () => {
  const tags = {
    name: 'thedude',
    hobby: 'bowling',
  };

  describe('generateAcmCertificate', () => {
    it('renders an acm certificate without tags', () => {
      const synthed = Testing.synthScope((stack) => {
        ApplicationCertificate.generateAcmCertificate(
          new TestResource(stack, 'test-resource'),
          'dev.gobowling.info'
        );
      });
      expect(synthed).toMatchSnapshot();
    });

    it('renders an acm certificate with tags', () => {
      const synthed = Testing.synthScope((stack) => {
        ApplicationCertificate.generateAcmCertificate(
          new TestResource(stack, 'test-resource'),
          'dev.gobowling.info',
          tags
        );
      });
      expect(synthed).toMatchSnapshot();
    });
  });

  describe('generateRoute53Record', () => {
    it('renders a route 53 record', () => {
      const synthed = Testing.synthScope((stack) => {
        const construct = new TestResource(stack, 'test-resource');

        const cert = ApplicationCertificate.generateAcmCertificate(
          construct,
          'dev.gobowling.info'
        );

        ApplicationCertificate.generateRoute53Record(
          construct,
          'dev.gobowling.info',
          cert
        );
      });
      expect(synthed).toMatchSnapshot();
    });
  });

  describe('generateAcmCertificateValidation', () => {
    it('renders an acm certificate validation', () => {
      const synthed = Testing.synthScope((stack) => {
        const construct = new TestResource(stack, 'test-resource');

        const cert = ApplicationCertificate.generateAcmCertificate(
          construct,
          'dev.gobowling.info'
        );

        const record = ApplicationCertificate.generateRoute53Record(
          construct,
          'dev.gobowling.info',
          cert
        );

        ApplicationCertificate.generateAcmCertificateValidation(
          construct,
          cert,
          record
        );
      });
      expect(synthed).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    it('throws an error without a zone id or zone domain', () => {
      const app = Testing.app();
      const stack = new TerraformStack(app, 'test');

      expect(() => {
        new ApplicationCertificate(stack, 'testCert', {
          domain: 'dev.gobowling.info',
        });
      }).toThrow('You need to pass either a zone id or a zone domain');
    });

    it('renders a cert with a zone id', () => {
      const synthed = Testing.synthScope((stack) => {
        new ApplicationCertificate(stack, 'testCert', {
          domain: 'dev.gobowling.info',
          zoneId: 'malibu',
        });
      });
      expect(synthed).toMatchSnapshot();
    });

    it('renders a cert with a zone domain', () => {
      const synthed = Testing.synthScope((stack) => {
        new ApplicationCertificate(stack, 'testCert', {
          domain: 'dev.gobowling.info',
          zoneId: 'gobowling.info',
        });
      });
      expect(synthed).toMatchSnapshot();
    });

    it('renders a cert with tags', () => {
      const synthed = Testing.synthScope((stack) => {
        new ApplicationCertificate(stack, 'testCert', {
          domain: 'dev.gobowling.info',
          zoneId: 'gobowling.info',
          tags: {
            name: 'thedude',
            hobby: 'bowling',
          },
        });
      });
      expect(synthed).toMatchSnapshot();
    });
  });
});
