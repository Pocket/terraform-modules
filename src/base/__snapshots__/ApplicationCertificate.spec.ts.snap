// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationCertificate constructor renders a cert with a zone domain 1`] = `
"{
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"testCert_certificate_3EDC39B5\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"validation_method\\": \\"DNS\\"
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"testCert_certificate_validation_80E8BDA3\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.testCert_certificate_3EDC39B5.arn}\\",
        \\"depends_on\\": [
          \\"aws_route53_record.testCert_certificate_record_93AF392D\\",
          \\"aws_acm_certificate.testCert_certificate_3EDC39B5\\"
        ],
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.testCert_certificate_record_93AF392D.fqdn}\\"
        ]
      }
    },
    \\"aws_route53_record\\": {
      \\"testCert_certificate_record_93AF392D\\": {
        \\"depends_on\\": [
          \\"aws_acm_certificate.testCert_certificate_3EDC39B5\\"
        ],
        \\"name\\": \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"gobowling.info\\"
      }
    }
  }
}"
`;

exports[`ApplicationCertificate constructor renders a cert with a zone id 1`] = `
"{
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"testCert_certificate_3EDC39B5\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"validation_method\\": \\"DNS\\"
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"testCert_certificate_validation_80E8BDA3\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.testCert_certificate_3EDC39B5.arn}\\",
        \\"depends_on\\": [
          \\"aws_route53_record.testCert_certificate_record_93AF392D\\",
          \\"aws_acm_certificate.testCert_certificate_3EDC39B5\\"
        ],
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.testCert_certificate_record_93AF392D.fqdn}\\"
        ]
      }
    },
    \\"aws_route53_record\\": {
      \\"testCert_certificate_record_93AF392D\\": {
        \\"depends_on\\": [
          \\"aws_acm_certificate.testCert_certificate_3EDC39B5\\"
        ],
        \\"name\\": \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"malibu\\"
      }
    }
  }
}"
`;

exports[`ApplicationCertificate constructor renders a cert with tags 1`] = `
"{
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"testCert_certificate_3EDC39B5\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"tags\\": {
          \\"hobby\\": \\"bowling\\",
          \\"name\\": \\"thedude\\"
        },
        \\"validation_method\\": \\"DNS\\"
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"testCert_certificate_validation_80E8BDA3\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.testCert_certificate_3EDC39B5.arn}\\",
        \\"depends_on\\": [
          \\"aws_route53_record.testCert_certificate_record_93AF392D\\",
          \\"aws_acm_certificate.testCert_certificate_3EDC39B5\\"
        ],
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.testCert_certificate_record_93AF392D.fqdn}\\"
        ]
      }
    },
    \\"aws_route53_record\\": {
      \\"testCert_certificate_record_93AF392D\\": {
        \\"depends_on\\": [
          \\"aws_acm_certificate.testCert_certificate_3EDC39B5\\"
        ],
        \\"name\\": \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.testCert_certificate_3EDC39B5.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"gobowling.info\\"
      }
    }
  }
}"
`;

exports[`ApplicationCertificate generateAcmCertificate renders an acm certificate with tags 1`] = `
"{
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test-resource_certificate_15680B70\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"tags\\": {
          \\"hobby\\": \\"bowling\\",
          \\"name\\": \\"thedude\\"
        },
        \\"validation_method\\": \\"DNS\\"
      }
    }
  }
}"
`;

exports[`ApplicationCertificate generateAcmCertificate renders an acm certificate without tags 1`] = `
"{
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test-resource_certificate_15680B70\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"validation_method\\": \\"DNS\\"
      }
    }
  }
}"
`;

exports[`ApplicationCertificate generateAcmCertificateValidation renders an acm certificate validation 1`] = `
"{
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test-resource_certificate_15680B70\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"validation_method\\": \\"DNS\\"
      }
    },
    \\"aws_acm_certificate_validation\\": {
      \\"test-resource_certificate_validation_7DB80AF0\\": {
        \\"certificate_arn\\": \\"\${aws_acm_certificate.test-resource_certificate_15680B70.arn}\\",
        \\"depends_on\\": [
          \\"aws_route53_record.test-resource_certificate_record_59EDF3C1\\",
          \\"aws_acm_certificate.test-resource_certificate_15680B70\\"
        ],
        \\"validation_record_fqdns\\": [
          \\"\${aws_route53_record.test-resource_certificate_record_59EDF3C1.fqdn}\\"
        ]
      }
    },
    \\"aws_route53_record\\": {
      \\"test-resource_certificate_record_59EDF3C1\\": {
        \\"depends_on\\": [
          \\"aws_acm_certificate.test-resource_certificate_15680B70\\"
        ],
        \\"name\\": \\"\${tolist(aws_acm_certificate.test-resource_certificate_15680B70.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.test-resource_certificate_15680B70.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.test-resource_certificate_15680B70.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"dev.gobowling.info\\"
      }
    }
  }
}"
`;

exports[`ApplicationCertificate generateRoute53Record renders a route 53 record 1`] = `
"{
  \\"resource\\": {
    \\"aws_acm_certificate\\": {
      \\"test-resource_certificate_15680B70\\": {
        \\"domain_name\\": \\"dev.gobowling.info\\",
        \\"lifecycle\\": {
          \\"create_before_destroy\\": true
        },
        \\"validation_method\\": \\"DNS\\"
      }
    },
    \\"aws_route53_record\\": {
      \\"test-resource_certificate_record_59EDF3C1\\": {
        \\"depends_on\\": [
          \\"aws_acm_certificate.test-resource_certificate_15680B70\\"
        ],
        \\"name\\": \\"\${tolist(aws_acm_certificate.test-resource_certificate_15680B70.domain_validation_options)[0].resource_record_name}\\",
        \\"records\\": [
          \\"\${tolist(aws_acm_certificate.test-resource_certificate_15680B70.domain_validation_options)[0].resource_record_value}\\"
        ],
        \\"ttl\\": 60,
        \\"type\\": \\"\${tolist(aws_acm_certificate.test-resource_certificate_15680B70.domain_validation_options)[0].resource_record_type}\\",
        \\"zone_id\\": \\"dev.gobowling.info\\"
      }
    }
  }
}"
`;
